pkg/notifier/beans/beans.go

```diff
@@ -0,0 +1,394 @@
+	}
+
+	return tempRequest
+}
+
+// GetIdsByTypeIndex
+// if new criteria fields are added, add a case to this function
+func (notificationConfigRequest *NotificationConfigRequest) GetIdsByTypeIndex(index selectorIndex) []*int {
+	switch index {
+	case teams:
+		return notificationConfigRequest.TeamId
+	case apps:
+		return notificationConfigRequest.AppId
+	case envs:
+		return notificationConfigRequest.EnvId
+	case clusters:
+		return notificationConfigRequest.ClusterId
+	}
+	return nil
+}
+
+// if new criteria fields are added, create a index for it here
+type selectorIndex int
+
+const teams, apps, envs, clusters selectorIndex = 0, 1, 2, 3
+
+// GenerateSettingCombinations
+// if new criteria is added , add a similar condition for that criteria index below
+func (notificationConfigRequest *NotificationConfigRequest) GenerateSettingCombinations() []*LocalRequest {
+	selectorIndices := make([]selectorIndex, 0)
+	if len(notificationConfigRequest.TeamId) > 0 {
+		selectorIndices = append(selectorIndices, teams)
+	}
+
+	if len(notificationConfigRequest.AppId) > 0 {
+		selectorIndices = append(selectorIndices, apps)
+	}
+
+	if len(notificationConfigRequest.EnvId) > 0 {
+		selectorIndices = append(selectorIndices, envs)
+	}
+
+	if len(notificationConfigRequest.ClusterId) > 0 {
+		selectorIndices = append(selectorIndices, clusters)
+	}
+
+	result := make([]*LocalRequest, 0)
+	if len(selectorIndices) == 0 {
+		return append(result, &LocalRequest{PipelineId: notificationConfigRequest.PipelineId})
+	}
+
+	selectorsMappedWithIdx := make([][]*int, 0)
+	for _, selectorIdx := range selectorIndices {
+		selectorsMappedWithIdx = append(selectorsMappedWithIdx, notificationConfigRequest.GetIdsByTypeIndex(selectorIdx))
+	}
+	generateCombinationSettings(selectorsMappedWithIdx, LocalRequest{}, 0, &result, selectorIndices)
+	return result
+}
+
+// example:
+// teams: [1,2]
+// apps: [3,4]
+// envs: [5,6,7]
+//
+//	we should generate the following combinations:
+//	(1,3,5), (1,3,6), (1,3,7),
+//	(1,4,5), (1,4,6), (1,4,7),
+//	(2,3,5), (2,3,6), (2,3,7),
+//	(2,4,5), (2,4,6), (2,4,7)
+//
+// these can be generated by just running  3 nested for loops. but whenever we use a new criteria lets say cluster,
+// we should add another for loop. this is also not a big challange, but the challenge is when we have random criterias ids, i.e  when we have
+// only (teamIds,appIds and clusterIds) or (teamIds,envIds and clusterIds) etc...
+// this requires us to select the combination of nested loops based on the available criteria.
+// this essentially requires 2^n combinations(n -> criteria selectors) of nested loop logic. to avoid this we are using the recursive approach.
+// generateCombinationSettings:
+// Function to generate all combinations of arrays corresponding to the selectorIndices
+func generateCombinationSettings(arrays [][]*int, current LocalRequest, index int, result *[]*LocalRequest, selectorIndices []selectorIndex) {
```

Comment:
this func seems like part of a util file

Reply:
since V1 function is defined here, we can keep this here itself